### PR TITLE
Issue 1724 - Porting logic of pre-handling minimatch patterns from azure-pipelines-task-lib

### PR DESCRIPTION
**Pull request by denis-tikhomirov**
Ported logic of pre-handling minimatch patterns from azure-pipelines-task-lib ([PS version](https://github.com/microsoft/azure-pipelines-task-lib/blob/21dac986b4c7478eb2ed8ffbd60e4fad62904438/powershell/VstsTaskSdk/FindFunctions.ps1#L306) / [TS version](https://github.com/microsoft/azure-pipelines-task-lib/blob/0bbb9f8f133f8052aa1cb602c2408470c1fe4afe/node/task.ts#L1406)). 
The logic allows to handle negative minimatch patterns correctly.
Also reporting of excluded items was moved from GetFilteredItems to GetArtifactItems.

Checked cases:
- download all (like `**`)
- download all (like `**`) except 1or 2 file (like `!**/file.1.nupkg`)
- download all (like `**`) except group of files (like `!**/file.*.nupkg`)
- download all (like `**`) except directory (like `!**/folder1/**`)
- download directory (like `**/folder1/**`) except 1 file (like `!**/folder1/file.1.nupkg`)
- download group files (like `**/file.*.nupkg`) except 1 file (like `!**/file.1.nupkg`)
- download all (like `**`) except directory (like `!**/folder1/**`) but with downloading 1 file from the directory (like `!!**/folder1/file.1.nupkg`)
- using of whitespaces in start and end of pattern
- using of sharp (`#`) in pattern
- using of empty pattern
- using of pattern without asterisks (full path to file in artifact)

Checklist:
- There’s no risky dependency updates
   - Corrected function is used by 2 tasks. And both these tasks work correctly with the changes.

- Changes has been tested
   - Yes. The changes are tested by 22 cases (2 affected tasks, 11 cases for each task) by build in prepared pipeline. Also test with 7 cases was added in these PR to cover corrected function.

- Enough test coverage for changes, and current test coverage for task doesn't look poor
   - Yes, test with 7 cases was added in these PR to cover corrected function. Most cases of using the function are covered.

- We understand how tasks is working, how changes affect task behavior
   - Yes. The changes allow to collect list of files which should be downloaded from artifact. Previously all files are downloaded.

- There is no breaking changes
   - No breaking changes.

- There is no any other concerns
   - No concerns. Corrected function is used by 2 tasks. And both these tasks work correctly with the changes.

- I have not discovered any new uncovered test/use cases
   - Yes, I don't uncovered test/use cases for corrected function.

### CODE REVIEW

**Comment by anatolybolshakov** [comment]
Please add tests to cover these changes.

**Comment by denis-tikhomirov** [comment]
> Please add tests to cover these changes.

Done. Tests were added to this PR.

File: src/Agent.Plugins/Artifact/FileContainerProvider.cs [hunk 1]
```diff
@@ -225,18 +231,32 @@ private async Task<IEnumerable<FileContainerItem>> GetArtifactItems(ArtifactDown
                 containerIdAndRoot.Item2
             );
 
-            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(
-                minimatchPatterns,
-                tracer,
-                downloadParameters.CustomMinimatchOptions
-            );
+            Options customMinimatchOptions;
+            if (!(downloadParameters.CustomMinimatchOptions is null))
```
Comment (EzzhevNikita) [inline src/Agent.Plugins/Artifact/FileContainerProvider.cs:235]: will be better to use this form `if (downloadParameters.CustomMinimatchOptions != null)`

File: src/Agent.Plugins/Artifact/FileContainerProvider.cs [hunk 2]
```diff
@@ -355,7 +375,158 @@ private string ResolveTargetPath(string rootPath, FileContainerItem item, string
             return absolutePath;
         }
 
-        private List<FileContainerItem> GetFilteredItems(List<FileContainerItem> items, IEnumerable<Func<string, bool>> minimatchFuncs)
+        public List<FileContainerItem> GetFilteredItems(List<FileContainerItem> items, string[] minimatchPatterns, Options customMinimatchOptions)
```
Comment (EzzhevNikita) [inline src/Agent.Plugins/Artifact/FileContainerProvider.cs:378]: Could we split this function into several ones?

File: src/Agent.Plugins/Artifact/FileContainerProvider.cs (hunk 1 above)
Comment (denis-tikhomirov) [reply src/Agent.Plugins/Artifact/FileContainerProvider.cs]: Corrected

File: src/Agent.Plugins/Artifact/FileContainerProvider.cs (hunk 2 above)
Comment (denis-tikhomirov) [reply src/Agent.Plugins/Artifact/FileContainerProvider.cs:378]: Done. Expanding of braces in patterns, updating of patterns map and applying patterns map to list of items were moved to separate functions.

**Comment by anatolybolshakov** [comment]
@denis-tikhomirov huge work, thanks! Let me take a look

**Review by tatyana-kostromskaya — APPROVED** [review]

**Review by EzzhevNikita — APPROVED** [review]
LGTM! Also tested changes locally, everything work as expected

**Review by anatolybolshakov — APPROVED** [review]

**Comment by denis-tikhomirov** [comment]
Changes from the PR will be split to 3 parts for following sub-tasks:
- porting logic of handling negative patterns for DownloadBuildArtifactsV1
- adding of tests for ported logic
- adding support of negative patterns to scenarios with using file share

**Comment by denis-tikhomirov** [comment]
Committed to master as:
- https://github.com/microsoft/azure-pipelines-agent/commit/7ea02035e8403ff9dda97dd314159eaff3d96b74
- https://github.com/microsoft/azure-pipelines-agent/commit/0c1261f003b1461ab65a149ad2ba9d42e6c05fa0
- https://github.com/microsoft/azure-pipelines-agent/commit/7c376535b6290be95f0082a3a45f95ba0f1c9e5f

**Pull request closed without merging**